zcash_client_sqlite/src/wallet.rs

```diff
@@ -656,47 +751,99 @@ pub(crate) fn get_current_address<P: consensus::Parameters>(
                     addr_str,
                 ))),
             })
-            .map(|addr| (addr, DiversifierIndex::from(di_be)))
+            .map(|addr| (addr, diversifier_index))
     })
     .transpose()
 }
 
-/// Adds the given address and diversifier index to the addresses table.
+/// Adds the given external address and diversifier index to the addresses table.
 ///
-/// Returns the database row for the newly-inserted address.
-pub(crate) fn insert_address<P: consensus::Parameters>(
+/// Returns the primary key identifier for the newly-inserted address.
+pub(crate) fn upsert_address<P: consensus::Parameters>(
     conn: &rusqlite::Connection,
     params: &P,
     account_id: AccountRef,
     diversifier_index: DiversifierIndex,
     address: &UnifiedAddress,
-) -> Result<(), SqliteClientError> {
+    exposed_at_height: Option<BlockHeight>,
+) -> Result<AddressRef, SqliteClientError> {
     let mut stmt = conn.prepare_cached(
         "INSERT INTO addresses (
             account_id,
             diversifier_index_be,
+            key_scope,
             address,
-            cached_transparent_receiver_address
+            transparent_child_index,
+            cached_transparent_receiver_address,
+            exposed_at_height
         )
         VALUES (
             :account_id,
             :diversifier_index_be,
+            :key_scope,
             :address,
-            :cached_transparent_receiver_address
-        )",
+            :transparent_child_index,
+            :cached_transparent_receiver_address,
+            :exposed_at_height
+        )
+        ON CONFLICT (account_id, diversifier_index_be, key_scope) DO UPDATE
```

Comment:
In the case that this upsert is triggered by a note that was received in a pool that the conflicting row's address does not contain, we should upgrade the row to contain that receiver (similar to what we do in `zcashd`), because this is evidence of that receiver being exposed to the ecosystem. It's not technically evidence of _other_ receivers in `address` being exposed, so we might only want to add the one receiver triggering the upsert to the conflicting row.